tensorflow_datasets/core/features/tensor_feature.py

```diff
@@ -131,19 +132,10 @@ def encode_example(self, example_data):
                        f'For {self}')
 
     np_dtype = np.dtype(self.dtype.as_numpy_dtype)
-    if isinstance(example_data, tf.Tensor):
-      raise TypeError(
-          f'Error encoding: {example_data!r}. `_generate_examples` should '
-          'yield `np.array` compatible values, not `tf.Tensor`')
-    if not isinstance(example_data, np.ndarray):
-      example_data = np.array(example_data, dtype=np_dtype)
-    # Ensure the shape and dtype match
-    if example_data.dtype != np_dtype:
-      raise ValueError('Dtype {} do not match {}'.format(
-          example_data.dtype, np_dtype))
-
-    shape = example_data.shape
-    utils.assert_shape_match(shape, self._shape)
+
+    example_data = np.asanyarray(example_data, dtype=np_dtype)
```

Comment:
`np.asanyarray` might accidentally cast data. It's best to be explicit and only call `asarray` if the data is not already a np.ndarray, as previously

Reply:
My understanding is that `asarray` and `asanyarray` behave identically in terms of data type casting. The only difference is that `asaanyarray` will let more types (e.g. `matrix`) to be converted without a copy.  This behavior seems okay since we only need a flat iterator.

Reply:
We do not want to cast the dtype.

If the user pass `np.array([.1, 0.4, .5])` but the expected dtype is int32, we want to raise an error. With this change, the float values will be silently casted to int32, instead of raising an error: `ValuError('Expected int32, got float32')`

Reply:
Hmm, but it is exactly what the old code does... it casts data type.

```python
if not isinstance(example_data, np.ndarray):
      example_data = np.array(example_data, dtype=np_dtype)
```

Do I read it incorrectly?

Reply:
> Hmm, but it is exactly what the old code does... it casts data type silently.

It will cast data if user pass `list[int]` (-> `np.array`), but not if the user pass a `np.array`